src/lib/blkalloc/append_blk_allocator.cpp

```diff
@@ -0,0 +1,162 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *   https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed
+ * under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR
+ * CONDITIONS OF ANY KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations under the License.
+ * *
+ * *********************************************************************************/
+#include <homestore/meta_service.hpp>
+#include "append_blk_allocator.h"
+
+namespace homestore {
+
+AppendBlkAllocator::AppendBlkAllocator(const BlkAllocConfig& cfg, bool first_time_boot, chunk_num_t id) :
+        BlkAllocator{cfg, id}, m_metrics{get_name().c_str()}, m_sb{get_name()} {
+    // all append_blk_allocator instances use same client type;
+    meta_service().register_handler(
+        get_name(),
+        [this](meta_blk* mblk, sisl::byte_view buf, size_t size) { on_meta_blk_found(std::move(buf), (void*)mblk); },
+        nullptr);
+
+    if (first_time_boot) {
+        m_sb.create(sizeof(append_blkalloc_sb));
+        m_sb->chunk_id = id;
+        m_sb->last_append_offset = 0;
+    }
+}
+
+void AppendBlkAllocator::on_meta_blk_found(const sisl::byte_view& buf, void* meta_cookie) {
+    m_sb.load(buf, meta_cookie);
+    HS_REL_ASSERT_EQ(m_sb->magic, append_blkalloc_sb_magic, "Invalid AppendBlkAlloc metablk, magic mismatch");
+    HS_REL_ASSERT_EQ(m_sb->version, append_blkalloc_sb_version, "Invalid version of AppendBlkAllocator metablk");
+}
+
+// alloc a single block;
+BlkAllocStatus AppendBlkAllocator::alloc(BlkId& bid) {
+    if (available_blks() < 1) {
+        COUNTER_INCREMENT(m_metrics, num_alloc_failure, 1);
+        LOGERROR("No space left to serve request nblks: 1, available_blks: {}", available_blks());
+        return BlkAllocStatus::SPACE_FULL;
+    }
+
+    bid.set(m_last_append_offset, 1, m_chunk_id);
+    ++m_last_append_offset;
+    --m_freeable_nblks;
+
+    set_dirty_offset();
+
+    COUNTER_INCREMENT(m_metrics, num_alloc, 1);
+    return BlkAllocStatus::SUCCESS;
+}
+
+//
+// For append blk allocator, the assumption is only one writer will append data on one chunk.
+// If we want to change above design, we can open this api for vector allocation;
+//
+BlkAllocStatus AppendBlkAllocator::alloc(blk_count_t nblks, const blk_alloc_hints& hint,
+                                         std::vector< BlkId >& out_bids) {
+    if (available_blks() < nblks) {
+        COUNTER_INCREMENT(m_metrics, num_alloc_failure, 1);
+        LOGERROR("No space left to serve request nblks: {}, available_blks: {}", nblks, available_blks());
+        return BlkAllocStatus::SPACE_FULL;
+    }
+
+    // Push 1 blk to the vector which has all the requested nblks;
+    out_bids.emplace_back(m_last_append_offset, nblks, m_chunk_id);
```

Comment:
sorry, why here 1 blk has all the requested nblks, but not n blks in the vector?

Reply:
because it is append only, nblks are always continuous and can be represented by 1 blkid.

Reply:
Next features we will remove the return of vector of blkids and return a CompositeBlkId where a single instance refers to multiple blks and all vdev calls accept this id for IO which internally runs multiple ios and get that. At present the vector of blkids are given to upper layer which needs to do lots of dance to persist them or serialize them. For Object store (using append blk allocator) this is never a problem because we return a single `BlkId`

Reply:
> it is append only, nblks are always continuous and can be represented by 1 blkid.

got it , but it is a little wired here to return a vector but having only one blk in it.

Reply:
@JacksonYao287 We will be removing the vector of BlkIds when we address this issue.
https://github.com/eBay/HomeStore/issues/151

Reply:
@JacksonYao287 For NuBlox, when there are not enough continuous blks that can serve the request, blk_allocator needs to go through chunks to collectively get all the blocks and put them in a vector. When there are enough continuous blocks, we only return 1 block_id which result in consumer to only do one write with the returned block id. Consumer can not assume it can always get 1 blkid for nblks that was requested. Hope this explains the doubt.